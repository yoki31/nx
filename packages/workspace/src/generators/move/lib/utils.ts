import {
  getWorkspaceLayout,
  joinPathFragments,
  normalizePath,
  ProjectConfiguration,
  Tree,
} from '@nx/devkit';
import { Schema } from '../schema';

/**
 * This helper function ensures that we don't move libs or apps
 * outside of the folders they should be in.
 *
 * This will break if someone isn't using the default libs/apps
 * folders. In that case, they're on their own :/
 */
export function getDestination(
  host: Tree,
  schema: Schema,
  project: ProjectConfiguration
): string {
  const projectType = project.projectType;

  const workspaceLayout = getWorkspaceLayout(host);

  let rootFolder = workspaceLayout.libsDir;
  if (projectType === 'application') {
    rootFolder = workspaceLayout.appsDir;
  }
  return joinPathFragments(rootFolder, schema.destination);
}

/**
 * Joins path segments replacing slashes with dashes
 *
 * @param path
 */
export function getNewProjectName(path: string): string {
  // strip leading '/' or './' or '../' and trailing '/' and replaces '/' with '-'
  return normalizePath(path)
    .replace(/(^\.{0,2}\/|\.{1,2}\/|\/$)/g, '')
    .split('/')
    .filter((x) => !!x)
    .join('-');
}

/**
 * Normalizes slashes (removes duplicates)
 *
 * @param input
 */
export function normalizePathSlashes(input: string): string {
  return (
    normalizePath(input)
      // strip leading ./ or /
      .replace(/^\.?\//, '')
      .split('/')
      .filter((x) => !!x)
      .join('/')
  );
}
