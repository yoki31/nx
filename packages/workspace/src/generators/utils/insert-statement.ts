import { applyChangesToString, ChangeType, Tree } from '@nx/devkit';
import { ensureTypescript } from '../../utilities/typescript';

let tsModule: typeof import('typescript');

/**
 * Insert a statement after the last import statement in a file
 */
export function insertStatement(tree: Tree, path: string, statement: string) {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }
  const { createSourceFile, isImportDeclaration, ScriptTarget } = tsModule;

  const contents = tree.read(path, 'utf-8');

  const sourceFile = createSourceFile(path, contents, ScriptTarget.ESNext);

  const importStatements = sourceFile.statements.filter(isImportDeclaration);
  const index =
    importStatements.length > 0
      ? importStatements[importStatements.length - 1].getEnd()
      : 0;

  if (importStatements.length > 0) {
    statement = `\n${statement}`;
  }

  const newContents = applyChangesToString(contents, [
    {
      type: ChangeType.Insert,
      index,
      text: statement,
    },
  ]);

  tree.write(path, newContents);
}
