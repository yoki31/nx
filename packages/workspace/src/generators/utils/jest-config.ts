import type { Tree } from '@nx/devkit';

export function findRootJestConfig(tree: Tree): string | null {
  if (tree.exists('jest.config.js')) {
    return 'jest.config.js';
  }

  if (tree.exists('jest.config.ts')) {
    return 'jest.config.ts';
  }

  return null;
}
