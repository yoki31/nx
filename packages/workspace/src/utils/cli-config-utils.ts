import type { Tree } from '@angular-devkit/schematics';

/**
 * @deprecated Nx no longer supports workspace.json
 */
export function getWorkspacePath(host: Tree) {
  const possibleFiles = ['/angular.json', '/workspace.json'];
  return possibleFiles.filter((path) => host.exists(path))[0];
}

export function parseTarget(targetString: string) {
  const [project, target, config] = targetString.split(':');
  return {
    project,
    target,
    config,
  };
}

export function editTarget(targetString: string, callback) {
  const parsedTarget = parseTarget(targetString);
  return serializeTarget(callback(parsedTarget));
}

/**
 * @deprecated use the utility from nx/src/utils instead
 */
export function serializeTarget({ project, target, config }) {
  return [project, target, config].filter((part) => !!part).join(':');
}
