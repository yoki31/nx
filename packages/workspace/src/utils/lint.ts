export const enum Linter {
  EsLint = 'eslint',
  None = 'none',
}
