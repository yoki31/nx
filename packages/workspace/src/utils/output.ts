import * as chalk from 'chalk';

export interface CLIErrorMessageConfig {
  title: string;
  bodyLines?: string[];
  slug?: string;
}

export interface CLIWarnMessageConfig {
  title: string;
  bodyLines?: string[];
  slug?: string;
}

export interface CLINoteMessageConfig {
  title: string;
  bodyLines?: string[];
}

export interface CLISuccessMessageConfig {
  title: string;
  bodyLines?: string[];
}

/**
 * Automatically disable styling applied by chalk if CI=true
 */
if (process.env.CI === 'true') {
  (chalk as any).level = 0;
}

class CLIOutput {
  private readonly NX_PREFIX = chalk.reset.inverse.bold.cyan(' NX ');
  /**
   * Longer dash character which forms more of a continuous line when place side to side
   * with itself, unlike the standard dash character
   */
  private readonly VERTICAL_SEPARATOR =
    '———————————————————————————————————————————————';

  /**
   * Expose some color and other utility functions so that other parts of the codebase that need
   * more fine-grained control of message bodies are still using a centralized
   * implementation.
   */
  colors = {
    gray: chalk.gray,
  };
  bold = chalk.bold;
  underline = chalk.underline;

  private writeToStdOut(str: string) {
    process.stdout.write(str);
  }

  private writeOutputTitle({
    label,
    title,
  }: {
    label?: string;
    title: string;
  }): void {
    let outputTitle: string;
    if (label) {
      outputTitle = `${this.NX_PREFIX} ${label} ${title}\n`;
    } else {
      outputTitle = `${this.NX_PREFIX} ${title}\n`;
    }
    this.writeToStdOut(outputTitle);
  }

  private writeOptionalOutputBody(bodyLines?: string[]): void {
    if (!bodyLines) {
      return;
    }
    this.addNewline();
    bodyLines.forEach((bodyLine) => this.writeToStdOut(`  ${bodyLine}\n`));
  }

  addNewline() {
    this.writeToStdOut('\n');
  }

  addVerticalSeparator() {
    this.writeToStdOut(`\n${chalk.gray(this.VERTICAL_SEPARATOR)}\n\n`);
  }

  addVerticalSeparatorWithoutNewLines() {
    this.writeToStdOut(`${chalk.gray(this.VERTICAL_SEPARATOR)}\n`);
  }

  error({ title, slug, bodyLines }: CLIErrorMessageConfig) {
    this.addNewline();

    this.writeOutputTitle({
      label: chalk.reset.inverse.bold.red(' ERROR '),
      title: chalk.bold.red(title),
    });

    this.writeOptionalOutputBody(bodyLines);

    /**
     * Optional slug to be used in an Nx error message redirect URL
     */
    if (slug && typeof slug === 'string') {
      this.addNewline();
      this.writeToStdOut(
        `${chalk.grey(
          '  Learn more about this error: '
        )}https://errors.nx.dev/${slug}\n`
      );
    }

    this.addNewline();
  }

  warn({ title, slug, bodyLines }: CLIWarnMessageConfig) {
    this.addNewline();

    this.writeOutputTitle({
      label: chalk.reset.inverse.bold.yellow(' WARNING '),
      title: chalk.bold.yellow(title),
    });

    this.writeOptionalOutputBody(bodyLines);

    /**
     * Optional slug to be used in an Nx warning message redirect URL
     */
    if (slug && typeof slug === 'string') {
      this.addNewline();
      this.writeToStdOut(
        `${chalk.grey(
          '  Learn more about this warning: '
        )}https://errors.nx.dev/${slug}\n`
      );
    }

    this.addNewline();
  }

  note({ title, bodyLines }: CLINoteMessageConfig) {
    this.addNewline();

    this.writeOutputTitle({
      label: chalk.reset.inverse.bold.keyword('orange')(' NOTE '),
      title: chalk.bold.keyword('orange')(title),
    });

    this.writeOptionalOutputBody(bodyLines);

    this.addNewline();
  }

  success({ title, bodyLines }: CLISuccessMessageConfig) {
    this.addNewline();

    this.writeOutputTitle({
      label: chalk.reset.inverse.bold.green(' SUCCESS '),
      title: chalk.bold.green(title),
    });

    this.writeOptionalOutputBody(bodyLines);

    this.addNewline();
  }

  logSingleLine(message: string) {
    this.addNewline();

    this.writeOutputTitle({
      title: message,
    });

    this.addNewline();
  }

  logCommand(message: string, isCached: boolean = false) {
    this.addNewline();

    this.writeToStdOut(chalk.bold(`> ${message} `));

    if (isCached) {
      this.writeToStdOut(chalk.bold.grey(`[retrieved from cache]`));
    }

    this.addNewline();
  }

  log({ title, bodyLines }: CLIWarnMessageConfig) {
    this.addNewline();

    this.writeOutputTitle({
      title: chalk.white(title),
    });

    this.writeOptionalOutputBody(bodyLines);

    this.addNewline();
  }
}

export const output = new CLIOutput();
