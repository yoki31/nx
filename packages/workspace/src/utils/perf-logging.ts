import { PerformanceObserver } from 'perf_hooks';

if (process.env.NX_PERF_LOGGING === 'true') {
  const obs = new PerformanceObserver((list) => {
    const entry = list.getEntries()[0];
    console.log(`Time for '${entry.name}'`, entry.duration);
  });
  obs.observe({ entryTypes: ['measure'] });
}
