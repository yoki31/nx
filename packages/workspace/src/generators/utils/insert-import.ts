import { insertStatement } from './insert-statement';
import { applyChangesToString, ChangeType, Tree } from '@nx/devkit';
import type { NamedImports } from 'typescript';
import { ensureTypescript } from '../../utilities/typescript';

let tsModule: typeof import('typescript');

export function insertImport(
  tree: Tree,
  path: string,
  name: string,
  modulePath: string
) {
  if (!tsModule) {
    tsModule = ensureTypescript();
  }
  const {
    createSourceFile,
    ScriptTarget,
    isStringLiteral,
    isImportDeclaration,
    isNamedImports,
  } = tsModule;

  const contents = tree.read(path, 'utf-8');

  const sourceFile = createSourceFile(path, contents, ScriptTarget.ESNext);

  const importStatements = sourceFile.statements.filter(isImportDeclaration);

  const existingImport = importStatements.find(
    (statement) =>
      isStringLiteral(statement.moduleSpecifier) &&
      statement.moduleSpecifier
        .getText(sourceFile)
        .replace(/['"`]/g, '')
        .trim() === modulePath &&
      statement.importClause.namedBindings &&
      isNamedImports(statement.importClause.namedBindings)
  );
  if (!existingImport) {
    insertStatement(tree, path, `import { ${name} } from '${modulePath}';`);
    return;
  }

  // TODO: Also check if the namedImport already exists
  const namedImports = existingImport.importClause
    .namedBindings as NamedImports;

  const index = namedImports.getEnd() - 1;

  let text: string;
  if (namedImports.elements.hasTrailingComma) {
    text = `${name},`;
  } else {
    text = `,${name}`;
  }

  const newContents = applyChangesToString(contents, [
    {
      type: ChangeType.Insert,
      index,
      text,
    },
  ]);

  tree.write(path, newContents);
}
