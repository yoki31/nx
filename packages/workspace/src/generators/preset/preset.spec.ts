import 'nx/src/internal-testing-utils/mock-project-graph';

import { readProjectConfiguration, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { presetGenerator } from './preset';
import { Preset } from '../utils/presets';

describe('preset', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' });
  });

  it(`should create files (preset = ${Preset.AngularMonorepo})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.AngularMonorepo,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.children('apps/proj')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/')).toMatchSnapshot();
    expect(tree.children('apps/proj/src/app')).toMatchSnapshot();
  }, 20000);

  it(`should create files (preset = ${Preset.WebComponents})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.WebComponents,
    });
    expect(tree.exists('/apps/proj/src/main.ts')).toBe(true);
  });

  it(`should create files (preset = ${Preset.ReactMonorepo})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactMonorepo,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.exists('/apps/proj/src/main.tsx')).toBe(true);
    expect(tree.read('apps/proj/webpack.config.js', 'utf-8')).toMatchSnapshot();
  });

  it(`should create files (preset = ${Preset.VueMonorepo})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.VueMonorepo,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.exists('apps/proj/src/main.ts')).toBe(true);
    expect(tree.read('apps/proj/vite.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it(`should create files (preset = ${Preset.Nuxt})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.Nuxt,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.exists('apps/proj/src/app.vue')).toBe(true);
    expect(readProjectConfiguration(tree, 'proj')).toBeDefined();
  });

  it(`should create files (preset = ${Preset.NextJs})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.NextJs,
      style: 'css',
      linter: 'eslint',
    });
    expect(tree.exists('/apps/proj/src/app/page.tsx')).toBe(true);
  });

  it(`should create files (preset = ${Preset.Express})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.Express,
      linter: 'eslint',
    });

    expect(tree.exists('apps/proj/src/main.ts')).toBe(true);
    expect(tree.exists('apps/proj/.eslintrc.json')).toBe(true);
  });

  it('should create files (preset = react-native)', async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactNative,
      linter: 'eslint',
    });

    expect(tree.exists('/apps/proj/src/app/App.tsx')).toBe(true);
  });

  it(`should create files (preset = ${Preset.ReactStandalone} bundler = webpack)`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactStandalone,
      style: 'css',
      linter: 'eslint',
      bundler: 'webpack',
    });
    expect(tree.exists('webpack.config.js')).toBe(true);
    expect(tree.read('webpack.config.js', 'utf-8')).toMatchSnapshot();
  });

  it(`should create files (preset = ${Preset.ReactStandalone} bundler = vite)`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.ReactStandalone,
      style: 'css',
      linter: 'eslint',
      bundler: 'vite',
    });
    expect(tree.exists('vite.config.ts')).toBe(true);
    expect(tree.read('vite.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it(`should create files (preset = ${Preset.VueStandalone})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.VueStandalone,
      style: 'css',
      e2eTestRunner: 'cypress',
    });
    expect(tree.exists('vite.config.ts')).toBe(true);
    expect(tree.read('vite.config.ts', 'utf-8')).toMatchSnapshot();
  });

  it(`should create files (preset = ${Preset.NuxtStandalone})`, async () => {
    await presetGenerator(tree, {
      name: 'proj',
      preset: Preset.NuxtStandalone,
      style: 'css',
      e2eTestRunner: 'cypress',
    });
    expect(tree.exists('nuxt.config.ts')).toBe(true);
    expect(readProjectConfiguration(tree, 'proj')).toBeDefined();
  });
});
