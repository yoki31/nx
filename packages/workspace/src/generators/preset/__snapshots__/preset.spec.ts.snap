// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preset should create files (preset = angular-monorepo) 1`] = `
[
  "project.json",
  "src",
  "tsconfig.app.json",
  "tsconfig.editor.json",
  "tsconfig.json",
  "public",
  ".eslintrc.json",
  "jest.config.ts",
  "tsconfig.spec.json",
]
`;

exports[`preset should create files (preset = angular-monorepo) 2`] = `
[
  "index.html",
  "styles.css",
  "app",
  "main.ts",
  "test-setup.ts",
]
`;

exports[`preset should create files (preset = angular-monorepo) 3`] = `
[
  "app.component.css",
  "app.component.html",
  "app.component.spec.ts",
  "app.component.ts",
  "app.module.ts",
  "nx-welcome.component.ts",
]
`;

exports[`preset should create files (preset = react-monorepo) 1`] = `
"const { NxAppWebpackPlugin } = require('@nx/webpack/app-plugin');
const { NxReactWebpackPlugin } = require('@nx/react/webpack-plugin');
const { join } = require('path');

module.exports = {
  output: {
    path: join(__dirname, '../../dist/apps/proj'),
  },
  devServer: {
    port: 4200,
  },
  plugins: [
    new NxAppWebpackPlugin({
      tsConfig: './tsconfig.app.json',
      compiler: 'babel',
      main: './src/main.tsx',
      index: './src/index.html',
      baseHref: '/',
      assets: ['./src/favicon.ico', './src/assets'],
      styles: ['./src/styles.css'],
      outputHashing: process.env['NODE_ENV'] === 'production' ? 'all' : 'none',
      optimization: process.env['NODE_ENV'] === 'production',
    }),
    new NxReactWebpackPlugin({
      // Uncomment this line if you don't want to use SVGR
      // See: https://react-svgr.com/
      // svgr: false
    }),
  ],
};
"
`;

exports[`preset should create files (preset = react-standalone bundler = vite) 1`] = `
"/// <reference types='vitest' />
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

export default defineConfig({
  root: __dirname,
  cacheDir: './node_modules/.vite/proj',

  server: {
    port: 4200,
    host: 'localhost',
  },

  preview: {
    port: 4300,
    host: 'localhost',
  },

  plugins: [react(), nxViteTsPaths()],

  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },

  build: {
    outDir: './dist/proj',
    emptyOutDir: true,
    reportCompressedSize: true,
    commonjsOptions: {
      transformMixedEsModules: true,
    },
  },

  test: {
    watch: false,
    globals: true,
    cache: {
      dir: './node_modules/.vitest/proj',
    },
    environment: 'jsdom',
    include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],

    reporters: ['default'],
    coverage: {
      reportsDirectory: './coverage/proj',
      provider: 'v8',
    },
  },
});
"
`;

exports[`preset should create files (preset = react-standalone bundler = webpack) 1`] = `
"const { NxAppWebpackPlugin } = require('@nx/webpack/app-plugin');
const { NxReactWebpackPlugin } = require('@nx/react/webpack-plugin');
const { join } = require('path');

module.exports = {
  output: {
    path: join(__dirname, './dist/proj'),
  },
  devServer: {
    port: 4200,
  },
  plugins: [
    new NxAppWebpackPlugin({
      tsConfig: './tsconfig.app.json',
      compiler: 'babel',
      main: './src/main.tsx',
      index: './src/index.html',
      baseHref: '/',
      assets: ['./src/favicon.ico', './src/assets'],
      styles: ['./src/styles.css'],
      outputHashing: process.env['NODE_ENV'] === 'production' ? 'all' : 'none',
      optimization: process.env['NODE_ENV'] === 'production',
    }),
    new NxReactWebpackPlugin({
      // Uncomment this line if you don't want to use SVGR
      // See: https://react-svgr.com/
      // svgr: false
    }),
  ],
};
"
`;

exports[`preset should create files (preset = vue-monorepo) 1`] = `
"/// <reference types='vitest' />
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

export default defineConfig({
  root: __dirname,
  cacheDir: '../../node_modules/.vite/apps/proj',

  server: {
    port: 4200,
    host: 'localhost',
  },

  preview: {
    port: 4300,
    host: 'localhost',
  },

  plugins: [vue(), nxViteTsPaths()],

  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },

  build: {
    outDir: '../../dist/apps/proj',
    emptyOutDir: true,
    reportCompressedSize: true,
    commonjsOptions: {
      transformMixedEsModules: true,
    },
  },

  test: {
    watch: false,
    globals: true,
    cache: {
      dir: '../../node_modules/.vitest/apps/proj',
    },
    environment: 'jsdom',
    include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],

    reporters: ['default'],
    coverage: {
      reportsDirectory: '../../coverage/apps/proj',
      provider: 'v8',
    },
  },
});
"
`;

exports[`preset should create files (preset = vue-standalone) 1`] = `
"/// <reference types='vitest' />
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';
import { nxViteTsPaths } from '@nx/vite/plugins/nx-tsconfig-paths.plugin';

export default defineConfig({
  root: __dirname,
  cacheDir: './node_modules/.vite/proj',

  server: {
    port: 4200,
    host: 'localhost',
  },

  preview: {
    port: 4300,
    host: 'localhost',
  },

  plugins: [vue(), nxViteTsPaths()],

  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },

  build: {
    outDir: './dist/proj',
    emptyOutDir: true,
    reportCompressedSize: true,
    commonjsOptions: {
      transformMixedEsModules: true,
    },
  },

  test: {
    watch: false,
    globals: true,
    cache: {
      dir: './node_modules/.vitest/proj',
    },
    environment: 'jsdom',
    include: ['src/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}'],

    reporters: ['default'],
    coverage: {
      reportsDirectory: './coverage/proj',
      provider: 'v8',
    },
  },
});
"
`;
